server/src/main/java/org/elasticsearch/action/bulk/IncrementalBulkService.java

```diff
@@ -14,74 +14,72 @@
 import org.elasticsearch.action.support.ActiveShardCount;
 import org.elasticsearch.client.internal.Client;
 import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
 import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
 import org.elasticsearch.common.util.concurrent.ThreadContext;
 import org.elasticsearch.core.Nullable;
 import org.elasticsearch.core.Releasable;
 import org.elasticsearch.core.Releasables;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.index.IndexingPressure;
-import org.elasticsearch.rest.action.document.RestBulkAction;
 
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Supplier;
 
+import static org.elasticsearch.common.settings.Setting.boolSetting;
+
 public class IncrementalBulkService {
 
+    public static final Setting<Boolean> INCREMENTAL_BULK = boolSetting(
+        "rest.incremental_bulk",
+        true,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
     private final Client client;
+    private final AtomicBoolean enabled = new AtomicBoolean(true);
     private final IndexingPressure indexingPressure;
     private final ThreadContext threadContext;
-    private final Supplier<Boolean> enabled;
 
     public IncrementalBulkService(Client client, IndexingPressure indexingPressure, ThreadContext threadContext) {
-        this(client, indexingPressure, threadContext, new Enabled());
-    }
-
-    public IncrementalBulkService(
-        Client client,
-        IndexingPressure indexingPressure,
-        ThreadContext threadContext,
-        ClusterSettings clusterSettings
-    ) {
-        this(client, indexingPressure, threadContext, new Enabled(clusterSettings));
-    }
-
-    public IncrementalBulkService(
-        Client client,
-        IndexingPressure indexingPressure,
-        ThreadContext threadContext,
-        Supplier<Boolean> enabled
-    ) {
         this.client = client;
         this.indexingPressure = indexingPressure;
         this.threadContext = threadContext;
-        this.enabled = enabled;
-    }
-
-    public boolean incrementalBulkEnabled() {
-        return enabled.get();
     }
 
     public Handler newBulkRequest() {
+        ensureEnabled();
         return newBulkRequest(null, null, null);
     }
 
     public Handler newBulkRequest(@Nullable String waitForActiveShards, @Nullable TimeValue timeout, @Nullable String refresh) {
+        ensureEnabled();
         return new Handler(client, threadContext, indexingPressure, waitForActiveShards, timeout, refresh);
     }
 
+    private void ensureEnabled() {
+        if (enabled.get() == false) {
+            throw new AssertionError("Unexpected incremental bulk request");
+        }
+    }
+
+    // This method only exists to tests that the feature flag works. Remove once we no longer need the flag.
```

Comment:
Can we put a similar comment on the `enabled` field (or rename it to `enabledForTests`)?